### ISSUE TITLE
How to export database based config from dev to staging to prod ?

### ISSUE THREAD

**Issue by storm1er** [strapi/strapi#672]
**Node.js version**:
v8.9.4
**npm version**:
5.6.0
**Strapi version**:
3.0.0-alpha.10.2
**Operating system**:
Ubuntu 17.10

**Do you want to request a *feature* or report a *bug*?**
Question

**What is the current behavior?**
Before alpha.10.1, I was able to easily upgrade from one env to another, especially when editing user-permission. But as you said, last upgrade moved all the config from file to database.
This ended up to create "one-shot-update" scripts, executed when deploying.
The point of 10- versions was that strapi handle this itself through files.

**What is the expected behavior?**
Maybe find some hybrid solution ?

* **Option 1 :** Keep configuration files, but can be overrided through ui
  * Should work in most case, but how do we handle file override an already overrided configuration ?
* **Option 2 :** Block user-permission edition to env, based on config
  * So now, the configuration will be project based : 
    * you choose to edit user-permission only from dev : file based
    * you choose to edit wherever you want user-permission : database config
  * It's probably the best answer but dev consuming (on strapi side)
* **Option 3 :** give us the possibility to have some "script on deploy" stuff that strapi could handle (using [this](https://strapi.io/documentation/configurations/configurations.html#database-configuration) ?)

**Comment by Aurelsicoko** [comment]
There are three possible solutions:

1 - Manually migrate the data from the table `users-permissions_permissions` development to production as you will have to do using others frameworks.
2 - Add a way to export permissions and import them in the production environment. But it's not optimal because you have to start the server (at this time, the permissions will be set with default configuration), then import the permissions. So you will have a lapse of time (few seconds) without the right permissions.
3 - The latest options (3) seems the best to me. It makes sense to me.

**Comment by storm1er** [comment]
@Aurelsicoko I agree with the latest option

**Comment by storm1er** [comment]
@lauriejim Can you transform this into "Feature request" ? ^^ And it's kinda important as arround me i'm not the only guy who is loosing their mind : "How to do this without destroy everything" x)

Most of us directly edit permission through mlab ^^'

**Comment by Aurelsicoko** [comment]
I would love to see someone try to make some work on this part. We need to improve it but we are too busy to start this work right now.

**Comment by storm1er** [comment]
Well ... to be fair, I think that moving from file to db was a bad move ...
I mean, moving ENTIRELY from file to db ... We should keep file, and on execution of `npm start` (eq. on each restart) the server reload what was saved from file.

Drupal is doing this and it's quite intuitive ... I'll look for more details if I have time.

**Comment by Aurelsicoko** [comment]
@storm1er However, if you update the permissions, you need to restart the server if they are stored in a file. That's the main reason it was better for us to move the permissions in the database.

**Comment by storm1er** [comment]
@Aurelsicoko No you don't need to : Settings are stored in database, and when needed, you "click upload permission" to recreate / regenerate permissions files.

In our case, when restarting the server, it wouldn't care about what was in database, it recreate permission through files (to build the database settings).

More details on this workflow example : 
[Drupal 8 - Configuration management](https://www.drupal.org/docs/8/configuration-management)  
[Drupal 8 - Managing your sites configuration](https://www.drupal.org/docs/8/configuration-management/managing-your-sites-configuration)

> By default, the "active" configuration is stored in the database ("config" table). This is for performance and security reasons. This is the complete configuration for the entire site at that moment. Configuration can be exported and imported as YAML files, either in its entirety, or a single piece of configuration, using Drush and/or Drupal Console config commands or the Configuration Manager. (See below for more details.)

Using this kind of workflow : 
- You can develop every detail of you website on your _develop_ environment
  - So you can sync every environment easily !
- You can deploy permission fix instead of manipulating strapi B.O. for every modifications
  - Think about business pipeline : dev/staging/qa/client/prod/... how many time will you lose ?
  - And edit database directly is always a bad idea, having something natively is always a better solution because you know (as a framework) what happens.
- You can still "quickfix" in prod when you missed something (through B.O.)
  - But yes, you have to recreate/regenerate your permission's file to keep them upon next deliveries
  - But no, you don't need to restart you server to apply what you edit through B.O.

What was great in strapi before that was the fact that EVERYTHING was versioned, whatever you were doing, just looking at the code was self-explanatory... Now we **must have** some deploy scripts to be sure that permissions are correctly set on every environment of our pipeline.

I supposed that not everyone will agree with this, but working with mates on the same project where you can't git permission (specially on API framework) feels really weird.

Sorry about writing mistakes, I don't often write so much english.

**Comment by storm1er** [comment]
@Aurelsicoko That's the idea !

About "_However, the risk is very high to forgot to export the permissions_"
You're right, but except in little project, most of the time it's highly unrecommended/prohibited to edit permission in prod, because we all know what happens when an intern have to edit something in prod :innocent: 

Also, a BIG :+1: about display a warning in the BO telling " :warning: permissions has been edited manually blablabla" & about splitting permissions.json :smiley_cat:

**Comment by storm1er** [comment]
Hi,

💯 ! Great work, my mind was near that idea but I couldn't find a way to make it "universal" for every strapi's components.

The only comment I have is; files are human-readable, which is perfect but,
- Why not use YAML files instead ? (not really important)
- Why using unix timestamp instead of date ? (kind of important, especially for rollback feature)

**Comment by Aurelsicoko** [comment]
We don't have any YAML file in the project. We are always using JSON files, it's easier to manipulate in JavaScript even if there are YAML to JSON converters.

Great idea to use a more readable date format 👍 

-----

As promised, a preview screenshot of the UI. We are also thinking about renaming the `Settings Manager` plugin to `Configurations` and make it accessible through the `Configurations` link in the `General` section in the left menu. It would make more sense, and we could integrate the **Migration** part into it.

![screen shot 2018-04-18 at 11 16 17](https://user-images.githubusercontent.com/4144726/38923046-ff1bd4c6-42f9-11e8-85e7-b4eac7488640.png)

**Comment by storm1er** [comment]
@lauriejim I see this has been deleted from 3.0.0-alpha.12.3 milestone, can we have some follow up on this ? Thanks :)

**Comment by Aurelsicoko** [comment]
@storm1er We're still focusing our efforts on the stability. Some of our resources aren't available as expected so we made the choice to delay the development of this feature.

**Comment by storm1er** [comment]
@Aurelsicoko Any news on the milestone targetted for this feature ? =)
Great jobs for the lasts release !

**Comment by Aurelsicoko** [comment]
Not at all... However, we merged a PR (https://github.com/strapi/strapi/pull/1514) which will certainly improve the transition between environment.

**Comment by derrickmehaffy** [comment]
@Aurelsicoko similar here, can probably close this in favor of: https://portal.productboard.com/strapi/c/12-migrate-across-environments

**Comment by boazpoolman** [comment]
Hi,

The past week I've been developing a new plugin which is very similar to the proposal of @Aurelsicoko above.
See [strapi-plugin-config-sync](https://github.com/boazpoolman/strapi-plugin-config-sync).

It is still a work in progress, and I'm very much open to suggestions on how to improve it.
If some of you could test it out and give some feedback, that would be much appreciated!

Happy sunday :)

**Comment by derrickmehaffy** [comment]
(Thank you for that plugin, super cool)

**Comment by BretCameron** [comment]
I had a similar issue to that described here, but with a more restricted goal - to make certain permissions public and persist those changes between environments. 

In case it's helpful to anyone, here's that plugin: [strapi-plugin-public-permissions](https://www.npmjs.com/package/strapi-plugin-public-permissions)

**Comment by andy250** [comment]
> In case it's helpful to anyone, here's that plugin: [strapi-plugin-public-permissions](https://www.npmjs.com/package/strapi-plugin-public-permissions)
 I have exact same requirement, I will give your plugin a try. 👍
